### PR TITLE
[Dist] enable to chunk node/edge data into arbitrary number of chunks 

### PR DESCRIPTION
**Pull request by Rhett-Ying**
## Description

for #4888 
## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: tools/distpartitioning/dataset_utils.py [hunk 1]
```diff
@@ -242,23 +243,19 @@ def get_dataset(input_dir, graph_name, rank, world_size, num_parts, schema_map):
...
-                        for idx in read_list[local_part_id]:
-                            feature_fname = feat_data[constants.STR_DATA][idx]
-                            if (os.path.isabs(feature_fname)):
-                                logging.info(f'Loading numpy from {feature_fname}')
-                                efeats.append(torch.from_numpy(np.load(feature_fname)))
-                            else:
-                                numpy_path = os.path.join(input_dir, feature_fname)
-                                logging.info(f'Loading numpy from {numpy_path}')
-                                efeats.append(torch.from_numpy(np.load(numpy_path)))
+                        start, end = edge_tids[etype_name][local_part_id]
```
Comment (classicsong) [inline tools/distpartitioning/dataset_utils.py:250]: Why not use the same loading method as before?  You have a read_list and each process load part of the files in a round-robin way.

Comment (Rhett-Ying) [reply tools/distpartitioning/dataset_utils.py:397]: read behavior is changed to previous one. and shuffle data is applied.

**Comment by Rhett-Ying** [comment]
@classicsong  Here's the data read logic.
https://github.com/Rhett-Ying/dgl/blob/c50ba39a5077a42f17dc4a16086330efd6af0b12/tools/distpartitioning/dataset_utils.py#L250-L264

**Review by jermainewang — APPROVED** [review]
Per our discussion, we will have keep the current `tools/chunk_graph.py` as it is for demonstration purposes while having another chunk graph utilities under the test folder. Otherwise, LGTM.

File: tools/distpartitioning/dataset_utils.py [hunk 2]
```diff
@@ -247,29 +383,53 @@ def get_dataset(input_dir, graph_name, rank, world_size, num_parts, schema_map):
...
-                num_chunks = len(feat_data[constants.STR_DATA])
-                read_list = np.array_split(np.arange(num_chunks), num_parts)
+                assert (feat_data[constants.STR_FORMAT][constants.STR_NAME]
+                    in [constants.STR_NUMPY, constants.STR_PARQUET])
+
+                edge_data = []
+                num_files = len(feat_data[constants.STR_DATA])
+                if num_files == 0:
+                    continue
+                read_list = np.array_split(np.arange(num_files), world_size)
```
Comment (classicsong) [inline tools/distpartitioning/dataset_utils.py:393]: Same question here

File: tools/distpartitioning/gloo_wrapper.py [hunk 3]
```diff
@@ -98,6 +111,8 @@ def alltoallv_cpu(rank, world_size, input_tensor_list):
         The size of the entire
     input_tensor_list : List of tensor
         The tensors to exchange
+    retain_nones : bool
+        Indicates whether to retain ``None`` data in returned value.
```
[the remark this answers is not in the document]
Comment (Rhett-Ying) [reply tools/distpartitioning/gloo_wrapper.py:115]: in some scenarios, we will receive nothing from remote peers and `None` will be assigned into corresponding position in output array. And we need retain such `None` values to keep the array size same as `world_size`. But sometimes, we just need to collect the `Non-None` data, so we don't retain `None` in the returned output.

File: tools/distpartitioning/dataset_utils.py [hunk 4]
```diff
@@ -1,14 +1,127 @@
...
+    data_shape = torch.tensor(data_shape, dtype=torch.int64)
+
+    data_shape_output = [
+        torch.zeros_like(data_shape) for _ in range(world_size)
+    ]
+    dist.all_gather(data_shape_output, data_shape)
+
+    # Rank~0 always succeeds to load non-empty data, so we fetch info from it.
+    data_dim = data_shape_output[0][1].item()
+    data_type = REV_DATA_TYPE_ID[data_shape_output[0][2].item()]
```
[the remark this answers is not in the document]
Comment (Rhett-Ying) [reply tools/distpartitioning/dataset_utils.py:80]: yes, right. So we need to gather the correct dtype from rank~0.

File: tools/distpartitioning/dataset_utils.py [hunk 5]
```diff
@@ -147,28 +260,51 @@ def get_dataset(input_dir, graph_name, rank, world_size, num_parts, schema_map):
...
-                # than num_partitions. 
-                num_chunks = len(feat_data[constants.STR_DATA])
-                read_list = np.array_split(np.arange(num_chunks), num_parts)
-                reader_fmt_meta = {"name": feat_data[constants.STR_FORMAT][constants.STR_NAME]}
+                # than num_partitions.
+                node_data = []
+                num_files = len(feat_data[constants.STR_DATA])
+                if num_files == 0:
+                    continue
+                read_list = np.array_split(np.arange(num_files), world_size)
```
[the remark this answers is not in the document]
Comment (Rhett-Ying) [reply tools/distpartitioning/dataset_utils.py:268]: Not round-robin, but deterministic if `num_files` and `world_size` is fixed. see example here:
```
>>> a
array([0, 1, 2, 3, 4])
>>> np.array_split(a, 3)
[array([0, 1]), array([2, 3]), array([4])]
```

File: tools/distpartitioning/dataset_utils.py (hunk 2 above)
Comment (Rhett-Ying) [reply tools/distpartitioning/dataset_utils.py:393]: Not round-robin, but deterministic if `num_files` and `world_size` is fixed

File: tools/distpartitioning/dataset_utils.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (classicsong) [reply tools/distpartitioning/dataset_utils.py:268]: Can you add a unitest for this logic? And if possible, can you provide some APIs under tools/distpartitioning, so users can call these APIs to decide which file to collect/download given the rank and the world_size.
Because our SageMaker solution will rely on this logic.

**Review by kylasa — APPROVED** [review]
Overall I agree with these changes. Please merge this PR, upon acceptance by M5 team.

File: tools/distpartitioning/dataset_utils.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Rhett-Ying) [reply tools/distpartitioning/dataset_utils.py]: added. pls check here: `//tools/distpartitioning/utils/generate_read_list`

File: tools/distpartitioning/utils.py [hunk 6]
```diff
@@ -557,3 +557,21 @@ def map_partid_rank(partid, world_size):
         id.
     """
     return partid % world_size
+
+
+def generate_read_list(num_files, world_size):
+    """Generate the file IDs to read for each rank.
```
Comment (classicsong) [inline tools/distpartitioning/utils.py:563]: Can you give an example in the docstring, as this API will be used by external users.

**Review by classicsong — APPROVED** [review]

File: tools/distpartitioning/utils.py (hunk 6 above)
Comment (Rhett-Ying) [reply tools/distpartitioning/utils.py:563]: Below is added.
```
Examples
    --------
    >>> tools.distpartitionning.utils.generate_read_list(10, 4)
    [array([0, 1, 2]), array([3, 4, 5]), array([6, 7]), array([8, 9])]
```
